### ISSUE TITLE
Ability to change FM stations

### ISSUE THREAD

**Issue by kklajbor** [mkulesh/onpc#16]
Hi, 
I have several stations defined, application does not see them - it would be nice to get this option.

**Comment by mkulesh** [comment]
Hi @kklajbor, thank you for issue. It is possible. However, my device does not supports FM; i.e I am not able to test anythings related to FM. Therefore, I did not yet implement any FM-specific messages within the communication layer. I see two ways to get this functionality working:
- If you could provide me the application logging for different situation I will be able implement all necessary FM-specific messages and expand the GUI. 
- If you already have any knowledge about Android development or wish to study it, we can design FM support together, I provide you all details about possible implementation and you implement this feature. 
That way do you prefer?

**Comment by kklajbor** [comment]
Hi @mkulesh , I programmed last time 20 years ago, so for now all I can offer is testing, logging and everything else :) - just let me know what I have to do. Have S9 on Android Pie, Onkyo is TX-NR656.

I am very interested in supporting your app, as official app from Onkyo started requesting gps access, which was a line crossing for me (gps for possibility to change volume or input? meh).

**Comment by mkulesh** [comment]
Hi @kklajbor, thanks! I created a wiki-page with instructions how to collect log: https://github.com/mkulesh/onpc/wiki/Collect-Logs
If you are able to perform these steps, please try the following scenario:
- Start the app
- Switch on the receiver
- In the app: If input selector is already set to FM, change it to NET
- Change input selector to FM
- Using standard remote control, start to play any station
- Make screenshots for MONITOR and MEDIA tabs
- Using standard remote control, select the next station (not from the list of saved stations)
- Make screenshots for MONITOR and MEDIA tabs again
- If you have any saved stations, select one of then using standard remote control (or standard Onkyo app)
- Make screenshots for MONITOR and MEDIA tabs again
- Switch off the receiver, close the app

If the official app from Onkyo requests such permissions (I agree, it is unnecessary permission), may be it is meaningful to post about it on HiFi forums  and let people known about existing alternatives like this app :)

**Comment by mkulesh** [comment]
Hi @kklajbor, do you have any progress with log?

**Comment by quelbs** [comment]
I just sent log files for FM and DAB as they are probably similar to each other in terms of protocol. I also had a look into the "Receiver information" and found out that it contains a <presetlist> part that contains the DAB and FM favourites.
```xml
<presetlist count="40">
<preset id="01" band="2" freq="0" name="some name " />
<preset id="02" band="2" freq="0" name="another na" />
...
<preset id="19" band="0" freq="0" name="" />
<preset id="1a" band="0" freq="0" name="" />
...
<preset id="1e" band="1" freq="89.00" name="FM 89.00  " />
<preset id="1f" band="1" freq="89.50" name="another fm" />
...
<preset id="27" band="0" freq="0" name="" />
<preset id="28" band="0" freq="0" name="" />
</presetlist>
```
I have DAB stations saved in the beginning ID 1 and following and some FM stations from ID 30 on. I changed the names of the stations, but probably they are required by DAB as there is no freq set...

**Comment by mkulesh** [comment]
Hi @quelbs, I started with this issue. Now, the FM presets are available in LISTEN tab. Unfortunately, I can not test it on my device. I expect that in the last nightly build the following features work:
- FM input channel can be selected in MEDIA tab, but the list of presets in the MEDIA tab is not filled (empty)
-  In the LISTEN tab, the "FM" label shall appear (see screenshot below)
- In the right of this label, the currently selected preset and total number of presets shall appear
- Below empty cover image, the name of the selected preset shall be visible
- Instead of playback buttons, two new buttons "Right" and "Left"  appear. If you press these buttons, current preset shall be changed

Could you please test how does it work?

![listen_fm](https://user-images.githubusercontent.com/30557348/54061708-4bc3c600-4202-11e9-89d6-daac6829569f.png)

**Comment by quelbs** [comment]
Most things are working as expected.

- [x] channel can be selected in MEDIA tab, but the list of presets in the MEDIA tab is not filled (empty)
- [x] In the LISTEN tab, the "FM" label shall appear (see screenshot below)
- [ ] In the right of this label, the currently selected preset and total number of presets shall appear
- [x] Below empty cover image, the name of the selected preset shall be visible
- [x] Instead of playback buttons, two new buttons "Right" and "Left" appear. If you press these buttons, current preset shall be changed

Either the total number should always be 40 (the number of list places available) or there should not be a second number at all, or the number of the current entry should be calculated. Currently  it is possible to have the first number larger than the second one which is confusing.
<img alt="Screenshot of new FM screen with preset active" src="https://user-images.githubusercontent.com/570222/54070218-81e26380-425d-11e9-8d38-c9b4614e9c81.jpg" width="250">

Also using the new "right" and "left" buttons switches from FM to DAB when the previous/next item in the list is a DAB station. (Yes, both types of stations are in one list!)

Missing feature: 
- [ ] changing the frequency<br>On the remote control pressing the up/down-keys are used for this, but the up/down-keys in the RC part of the APP do not work for me here.

**Comment by mkulesh** [comment]
Hi @quelbs

> Also using the new "right" and "left" buttons switches from FM to DAB when the previous/next item in the list is a DAB station. (Yes, both types of stations are in one list!)

If channel is switched from FM to DAB, does the media input indication in the MEDIA tab change from FM to DAB  as well? If yes, I think it is a good feature :)

**Comment by quelbs** [comment]
Yes, this is changed, too. But in DAB view in LISTEN Tab, the new preset switch buttons are not available, thus i cannot check the other way round.

**Comment by mkulesh** [comment]
Hi @quelbs,
that do you  expect to see in MEDIA tab? All FM presets for FM media input channel and all DAB presets for DAB channel? 

> Yes, this is changed, too. But in DAB view in LISTEN Tab, the new preset switch buttons are not available, thus i cannot check the other way round.
I will commit shortly a change that also allows AM and DAB too

**Comment by quelbs** [comment]
I think this would be nice, an additional indicator what type of entry it is would be nice (similar to the indicators in the NET tab)

**Comment by mkulesh** [comment]
Hi @quelbs, I am now committed the visibility of presets in the MEDIA tab:
- In the LISTEN tab, DAB is now available
- For the FM input selector, only FM presets shall be visible
- For the DAB input selector, only DAB presets shall be visible
- The station that is currently playing shall have "Play" icon with accent color, see screenshot below
- I expect that is you press on any preset the station shall be changed
Does it work?

![media_fm](https://user-images.githubusercontent.com/30557348/54071448-de00b400-426c-11e9-9c46-3dbb40aab651.png)

**Comment by mkulesh** [comment]
Hi @quelbs, the frequency tuning is now possible:
-  there are added two new buttons "Fast forward" and "Fast backwards" in the LISTEN tab. These buttons increase and decrease the frequency. 
- Current frequency is now shown in the top and right of empty cover image, see screenshot above. I expect that this frequency shall be changed if you click on the "Fast forward" and "Fast backwards" buttons.  
- I changes the indication in the top and left of the empty cover image: now it always shows the total number of all presets (40). Is it OK or do you think there is a better variant? 

![media_fm](https://user-images.githubusercontent.com/30557348/54073936-2334de00-428d-11e9-8ce5-63c23b34545b.png)

**Comment by quelbs** [comment]
Thank you for the major improvements. Everything looks much better now. The only problems I figured out:
- If you listened to a media that has more information than just the station name and switch to DAB, the information of the previous title are not changed, also the times and slider position are not reset.
- If you restart the app the Name/Freq is not shown in LISTEN tab and not highlighted in MEDIA Tab

I don't know if this one is a problem, but I switched tabs between LISTEN and MEDIA multiple times (without selecting anything) and the playback stopped and the Receiver seemed to hang for some time also the display on the device lost the name of the station and the receiver was not changing channel any longer. The problem was persistent even when going in standby and powering on again afterwards. Only unplugging the Receiver helped to recover.

**Comment by mkulesh** [comment]
Hi @quelbs, I fixed these two problems. 
 
> I don't know if this one is a problem, but I switched tabs between LISTEN and MEDIA multiple times (without selecting anything) and the playback stopped and the Receiver seemed to hang for some time also the display on the device lost the name of the station and the receiver was not changing channel any longer.

Is this behavior reproducible? If yes, could you please collect and send me the app logging for this situation?

**Comment by quelbs** [comment]
Hi @mkulesh 
The behaviour was not reproducible, the other problems are solved for me.

**Comment by mkulesh** [comment]
Hi @quelbs, could you please send me one or two screenshots with FM and DAB? I would like to add them to the images library on https://github.com/mkulesh/onpc/tree/master/images/screenshots

**Comment by mkulesh** [comment]
Hi @kklajbor, the issue is implemented. It would be nice if you could test it wit the latest nightly build: https://github.com/mkulesh/onpc/raw/autobuild/autobuild/onpc-v0.9-debug.apk

**Comment by mkulesh** [comment]
Released in version 0.9
